sklearn/neighbors/_graph.py

```diff
@@ -36,6 +43,20 @@ def _query_include_self(X, include_self, mode):
     return X
 
 
+@validate_params(
+    {
+        "X": ["array-like", BallTree, KNeighborsMixin],
```

Comment:
Oh yes, I see. Soo your solution of using a `KNeighborsMixin` is probably the safest then. We only impose third-party libraries to inherit from this mixin class.

Reply:
It was @adrinjalali's suggestion. But don't we need to include RadiusNeighborsMixin as well?

Reply:
Since the function is specialized for k-neighbors (`kneighbors_graph`) then I don't think it makes sense to add `RadiusNeighborsMixin` here.

Reply:
Hm, sorry to ask again, but `kneighbors_graph` is currently functioning with `RadiusNeighborsClassifier` and `RadiusNeighborsTransformer` as inputs for the `X` param, even though it might not have been intended to be used that way. (I believe there is  `radius_neighbors_graph` for that purpose.)
So, if we now validate the param X not to be any of the radius neighbors estimators anymore, somebodie's code could break, correct?

Reply:
> So, if we now validate the param X not to be any of the radius neighbors estimators anymore, somebodie's code could break, correct?

I have to think a bit more about it to know if this is a normal use case or if this is just a bug. If this is a feature then we need to be lenient; if this is a bug then we can break people's code and add an entry in the changelog.

But I have to check more the details of how this function is used ;)

Reply:
> I don't know if we should check for KNeighborsMixin or just an estimator that defines a kneighbors_graph function 

the implementation explicitly checks for a `KNeighborsMixin`. I don't even think a BallTree makes sense here. The 2 valid options are an array or a fitted neighbors model. To me the docstring is wrong and needs to be fixed